x-pack/plugins/observability_solution/logs_explorer/public/state_machines/logs_explorer_controller/src/services/data_view_service.ts

```diff
@@ -26,3 +27,23 @@ export const createAndSetDataView =
      */
     discoverStateContainer.actions.setDataView(dataView);
   };
+
+export const changeDataView =
+  ({
+    dataViews,
+  }: {
+    dataViews: DataViewsPublicPluginStart;
+  }): InvokeCreator<LogsExplorerControllerContext, LogsExplorerControllerEvent> =>
+  async (context) => {
+    if (!('discoverStateContainer' in context)) return;
+    const { discoverStateContainer } = context;
+
+    // We need to manually retrieve the data view and force a set and change on the state container
+    // to guarantee the correct update on the data view selection and avoid a race condition
+    // when updating the control panels.
+    const nextDataView = await dataViews.get(context.datasetSelection.toDataviewSpec().id, false);
+    discoverStateContainer.actions.setDataView(nextDataView);
+    if (nextDataView.id) {
+      await discoverStateContainer.actions.onChangeDataView(nextDataView.id);
```

Comment:
Can you help me understand why we need this here but not in the ad-hoc case?

Reply:
Sure, we are updating the data views in this way as follows due to the same reason explained on the first comment above the `discoverStateContainer.actions.setDataView(dataView)` usage.
The difference here is the following:

### Create ad-hoc data view starting from a dataset
1. invoke `discoverStateContainer.actions.createAndAppendAdHocDataView`
  1.1 `createAndAppendAdHocDataView` creates a new ad-hoc data view
  1.2 invokes the `onChangeDataView` internally to change data view
2. set manually the newly created ad-hoc data view to avoid the race condition

### Switch to existing data view
1. Retrieve and change existing data view
  1.1 Retrieve data view by id using the dataViews service (discoverStateContainer does not expose an action to retrieve singular data views)
  1.2 invokes the `onChangeDataView` manually to change data view as happened for ad-hoc case
2. set manually the retrieved data view to avoid the race condition

I feel this is still hacky and a patch to the state updates internally triggered on Discover, but I couldn't find yet a better way to fix it.
I hope the steps above clarify what happens here, let me know if anything is not clear

Reply:
ok, thanks for the explanation